// @sudoku/Game_pack.js

import { Sudoku } from './sudoku_pack.js';

// 纯领域 Game 类，无 UI/store 依赖
export class Game {
    #sudoku;
    #started;
    #history;
    #currentIndex;

    constructor(sudoku) {
        this.#started = false; 
        this.#history = [];
        this.#currentIndex = -1;
        this.#sudoku = new Sudoku(sudoku) || new Sudoku();
        this.saveHistory(); // 初始化时保存初始状态
    }

    get started() {
        return this.#started;
    }

    undo() {
        if (this.#currentIndex <= 0) return;
        this.#currentIndex--;
        this.#sudoku.fromJSON(this.#history[this.#currentIndex]);
    }

    redo() {
        if (this.#currentIndex >= this.#history.length - 1) return;
        this.#currentIndex++;
        const nextState = this.#history[this.#currentIndex];
        this.#sudoku.fromJSON(nextState);
    }

    saveHistory() {
        // 如果当前在撤销中间，新操作会清除未来的重做历史
        if (this.#currentIndex < this.#history.length - 1) {
            this.#history = this.#history.slice(0, this.#currentIndex + 1);
        }
        // 保存数独当前状态
        const state = this.#sudoku.toJSON();
        this.#history.push(state);
        this.#currentIndex = this.#history.length - 1;
    }

    getSudoku() {
        return this.#sudoku;
    }

    guess(move) {
        const success = this.#sudoku.guess(move);
        if (!success) return false;
        this.saveHistory();
        return true;
    }

    canUndo() {
        return this.#currentIndex > 0;
    }

    canRedo() {
        return this.#currentIndex < this.#history.length - 1;
    }

    toJSON() {
        return {
            sudoku: this.#sudoku.toJSON(),
            history: this.#history,
            currentIndex: this.#currentIndex
        };
    }
    

    static fromJSON(json) {
        const game = new Game();
        game.fromJSON(json);
        return game;
    }

    fromJSON(json) {
        if (!json || !json.sudoku || !json.sudoku.grid || !this.#sudoku.isValidGrid(json.sudoku.grid)) {
			throw new Error('Invalid JSON');
		}
        this.#sudoku.fromJSON(json.sudoku);
        this.#history = [...json.history];
        this.#currentIndex = json.currentIndex;
        this.#started = true;
    }
 isValidGameJSON(json) {
    if (typeof json !== 'object' || json === null) return false;
    if (!json.sudoku || !json.sudoku.grid) return false;
    if (!this.#sudoku.isValidGrid(json.sudoku.grid)) return false;
    if (!Array.isArray(json.history)) return false;
    if (typeof json.currentIndex !== 'number' || json.currentIndex < 0 || json.currentIndex >= json.history.length) return false;
    // 校验 history 里的每个快照
    for (const snap of json.history) {
        if (!snap.grid || !this.#sudoku.isValidGrid(snap.grid)) return false;
    }
    return true;
}
   
}
