// @sudoku/gamestore.js
import { writable, get, derived } from 'svelte/store';
import { Game } from './Game_pack';
import { Sudoku } from './sudoku_pack';
import { generateSudoku, solveSudoku } from '@sudoku/sudoku';
import { decodeSencode, encodeSudoku } from '@sudoku/sencode';
import { hints } from './stores/hints';

function createEmptyGame() {
	const emptyGrid = Array.from({ length: 9 }, () => Array(9).fill(0));
	return new Game(emptyGrid);
}

function createGameStore() {
	const { subscribe, set, update } = writable(createEmptyGame());

	return {
		subscribe,

		
		startNew(difficulty) {
			const grid = generateSudoku(difficulty);
			set(new Game(grid));
		},

		startCustom(sencode) {
			const grid = decodeSencode(sencode);
			set(new Game(grid));
		},

		startCreatorMode() {
			set(createEmptyGame());
		},

		get(pos) {
			const game = get(this);
			if (!game) return 0;
			return game.getSudoku().getGrid()[pos.y][pos.x];
		},

		getSencode() {
			const game = get(this);
			if (!game) return '';
			return encodeSudoku(game.getSudoku().getGrid());
		},

		set(pos, value) {
			update(game => {
				game?.guess({ row: pos.y, col: pos.x, value });
				return game;
			});
		},

		applyHint(pos) {
			hints.useHint();
			const game = get(this);
			if (!game) return;
			const solved = solveSudoku(game.getSudoku().getGrid());
			update(g => {
				g?.guess({
					row: pos.y,
					col: pos.x,
					value: solved[pos.y][pos.x]
				});
				return g;
			});
		},

		
		undo() {
			update(game => { game?.undo(); return game; });
		},

		redo() {
			update(game => { game?.redo(); return game; });
		},

	
		getGrid() {
			const game = get(this);
			return game ? game.getSudoku().getGrid() : Array(9).fill(Array(9).fill(0));
		},

		isWon() {
			const game = get(this);
			return game ? game.getSudoku().isSolved() : false;
		}
	};
}

export const gameStore = createGameStore();
export const gamePaused = writable(false);
export const gameWon = derived(
    gameStore,
    ($gameStore) => {
        if (!$gameStore) return false;
        return $gameStore.getSudoku().isSolved();
    },
    false
);
