// @sudoku/sudoku_pack.js

import {  SUDOKU_SIZE } from './constants';



export class Sudoku {
	#grid;
	#initGrid;

        constructor(grid) {
            if (grid && !this.isValidGrid(grid)) {
                throw new Error('Invalid grid');
            }
            if (Array.isArray(grid)) { 
                // 用已有 grid 初始化
                this.#grid = grid.map(row => [...row]);
				this.#initGrid = grid.map(row => [...row]);
            } else {
                // 默认空盘面
                this.#grid = Array.from({ length: SUDOKU_SIZE }, () => Array(SUDOKU_SIZE).fill(0));
            }
           
        }
        changeGrid(newGrid) {
            if (!this.isValidGrid(newGrid)) {
                throw new Error('Invalid grid');
            }
            for (let y = 0; y < SUDOKU_SIZE; y++) {
                for (let x = 0; x < SUDOKU_SIZE; x++) {
                    this.#grid[y][x] = newGrid[y][x];
                }
            }
            this.#initGrid = newGrid.map(row => [...row]);
        }
        getGrid() {
            // 返回当前盘面（深拷贝）
            return this.#grid.map(row => [...row]);
        }

        guess(move) {
            const { row, col, value } = move;
            const temp = this.#grid[row][col];
            if (this.#initGrid[row][col] === 0) {
                this.#grid[row][col] = value;
            } else {
				return false; // 不能修改已填数字
                
            }
            if (!this.isValidGrid(this.#grid)) {
                this.#grid[row][col] = temp;
			return false;}
            return true;
        }

        clone() {
            const cloned = new Sudoku(this.getGrid());
            cloned.#initGrid = this.#initGrid.map(row => [...row]);
            return cloned;
        }

        toJSON() {
            return {
                grid: this.getGrid(),
				initGrid: this.#initGrid.map(row => [...row])
            };
        }

	 fromJSON(json) {
    if (!json || !json.grid || !this.isValidGrid(json.grid)) throw new Error('Invalid JSON');
    this.#grid = json.grid.map(row => [...row]);
    // 如果有 initGrid 也要恢复
    if (json.initGrid) this.#initGrid = json.initGrid.map(row => [...row]);
    }

        static fromJSON(json) {

            const checker = new Sudoku();
            if (!json || !json.grid || !checker.isValidGrid(json.grid)) throw new Error('Invalid JSON');
            return new Sudoku(json.grid);
        }

        toString() {
            const grid = this.getGrid();
            return grid.map(row => row.map(num => num === 0 ? '.' : num).join(' ')).join('\n');
        }
        isValidGrid(grid) {
             if (!Array.isArray(grid) || grid.length !== 9) return false;
             for (const row of grid) {
             if (!Array.isArray(row) || row.length !== 9) return false;
             for (const cell of row) {
                if (!Number.isInteger(cell) || cell < 0 || cell > 9) return false;
        }
    }
    // 检查每行、每列、每宫无重复（忽略0）
    for (let i = 0; i < 9; i++) {
        const rowSet = new Set();
        const colSet = new Set();
        for (let j = 0; j < 9; j++) {
            // 行
            const r = grid[i][j];
            if (r !== 0) {
                if (rowSet.has(r)) return false;
                rowSet.add(r);
            }
            // 列
            const c = grid[j][i];
            if (c !== 0) {
                if (colSet.has(c)) return false;
                colSet.add(c);
            }
        }
    }
    // 检查每个3x3宫
    for (let boxRow = 0; boxRow < 3; boxRow++) {
        for (let boxCol = 0; boxCol < 3; boxCol++) {
            const boxSet = new Set();
            for (let i = 0; i < 3; i++) {
                for (let j = 0; j < 3; j++) {
                    const val = grid[boxRow * 3 + i][boxCol * 3 + j];
                    if (val !== 0) {
                        if (boxSet.has(val)) return false;
                        boxSet.add(val);
                    }
                }
            }
        }
    }
    return true;
}
isSolved() {
    // 1. 检查是否有空格
    for (let row = 0; row < SUDOKU_SIZE; row++) {
        for (let col = 0; col < SUDOKU_SIZE; col++) {
            if (this.#grid[row][col] === 0) return false;
        }
    }
    // 2. 检查当前盘面是否合法
    return this.isValidGrid(this.#grid);
}
        
    }
